src/Google.Events.Protobuf/ProtobufJsonCloudEventFormatter.cs

```diff
@@ -0,0 +1,101 @@
+        public override void DecodeBinaryModeEventData(byte[] data, CloudEvent cloudEvent) =>
+            cloudEvent.Data = s_jsonParser.Parse<T>(new StreamReader(new MemoryStream(data)));
+
+        // TODO: Handle other structured modes beyond JSON?
+
+        /// <inheritdoc />
+        public override CloudEvent DecodeStructuredModeMessage(byte[] data, ContentType contentType, IEnumerable<CloudEventAttribute> extensionAttributes) =>
+            s_structuredEventFormatter.DecodeStructuredModeMessage(data, contentType, extensionAttributes);
+
+        /// <inheritdoc />
+        public override IReadOnlyList<CloudEvent> DecodeBatchModeMessage(byte[] data, ContentType contentType, IEnumerable<CloudEventAttribute> extensionAttributes) =>
+            s_structuredEventFormatter.DecodeBatchModeMessage(data, contentType, extensionAttributes);
+
+        /// <inheritdoc />
+        public override byte[] EncodeBinaryModeEventData(CloudEvent cloudEvent)
+        {
+            var data = (T)cloudEvent.Data;
+            if (data is null)
+            {
+                return Array.Empty<byte>();
+            }
+            var json = data.ToString();
+            return Encoding.UTF8.GetBytes(json);
+        }
+
+        /// <inheritdoc />
+        public override byte[] EncodeStructuredModeMessage(CloudEvent cloudEvent, out ContentType contentType) =>
+            s_structuredEventFormatter.EncodeStructuredModeMessage(cloudEvent, out contentType);
+
+        /// <inheritdoc />
+        public override byte[] EncodeBatchModeMessage(IEnumerable<CloudEvent> cloudEvents, out ContentType contentType) =>
+            s_structuredEventFormatter.EncodeBatchModeMessage(cloudEvents, out contentType);
+
+        private class StructuredEventFormatter : JsonEventFormatter
+        {
+            protected override void EncodeStructuredModeData(CloudEvent cloudEvent, Utf8JsonWriter writer)
+            {
+                writer.WritePropertyName("data");
+                // TODO: Try to make this cleaner. It's annoying that we have to convert to a JSON string,
+                // reparse, then reserialize.
+                using var document = JsonDocument.Parse(cloudEvent.Data.ToString());
+                document.WriteTo(writer);
+            }
+
+            protected override void DecodeStructuredModeDataProperty(JsonElement dataElement, CloudEvent cloudEvent)
+            {
+                if (dataElement.ValueKind != JsonValueKind.Object)
+                {
+                    throw new InvalidOperationException("Expected Data property to have an object value");
+                }
+                string json = dataElement.GetRawText();
+                cloudEvent.Data = s_jsonParser.Parse<T>(json);
+            }
+
+            protected override void DecodeStructuredModeDataBase64Property(JsonElement dataBase64Element, CloudEvent cloudEvent)
+            {
+                throw new InvalidOperationException("Expected Data property to be set");
```

Comment:
Could you accept here the base64 representation of the binary encoded proto? And maybe emit that as well one method up depending on the CloudEvent.DataContentType?
Would that be useful? I'm not sure.

Reply:
We *could( potentially do that in the future, yes - although I *suspect* we'd want to just use the protobuf event format (which is a whole other thing) instead.

(I originally optionally used the binary representation in EncodeBinaryModeEventData based on the content type, then decided not to - it's too forward-looking with no real reason to expect that that will happen.)

Reply:
OK, yep, postponing for if/when is needed is totally fine.